### DISCUSSION TITLE
How to make deploy dependent on linting and unit tests

### DISCUSSION THREAD

**Discussion by NickR49** [vercel/community#845] [Help] [answered]
Hiya,

To date I've been running linting and unit tests within my Vercel NextJS deploys by running `pnpm build-test` as my build command which in turn runs `jest && next lint && next build` (so build conveniently fails if linting or tests don't past). However I've enabled my React Testing Library tests which expects the node environment to not be `production`. Now of course I want to keep my environment as production for the build. Is there a way to move the linting check and unit tests check out of the build step? Perhaps there's an integration for that?

I could use `NODE_ENV='test' jest && next lint && NODE_ENV='production' next build`, but it does feel a bit funny squeezing my tests into the build step.

TIA!
Nick

**Comment by mehulkar** [comment, marked as answer]
It's a good idea to have separate scripts for testing before deploy, especially if you want to set environment variables differently for each. There are many ways to do that. 

Outside of `turbo`, you could use [npm `pre` scripts](https://docs.npmjs.com/cli/v8/using-npm/scripts#pre--post-scripts) for this. E.g. in package.json:

```jsonc
// package.json
{
  "scripts": {
    "predeploy": "NODE_ENV='test' jest && next lint",
    "deploy": "NODE_ENV='production' next build"
  }
}
```

If you want to take advantage of Turbo's caching and parallelization, you may instead want to design your test/lint scripts _without_ using pre scripts:


```jsonc
// package.json
{
  "scripts": {
    "test": "NODE_ENV='test' jest && next lint",
    "deploy": "NODE_ENV='production' next build"
  }
}
```

and then tell turbo that deploy depends on test:

```jsonc
{
  "pipeline": {
    "deploy": {
      // all "deploy" tasks depend on a "test" task in their _own_ workspace
      "dependsOn": ["test"]
    }
  }
}
```

Note that this turbo config means something more than just "run this command before this other one", so be sure to [read the docs](https://turborepo.org/docs/core-concepts/running-tasks)! Hope this helps.

**Reply by NickR49** [reply]
NPM pre-scripts do the trick - thanks!
